BigTreeTech CB2/Pi2 config

JohnTheCoolingFan opened the pull request:
# Description

Added patches and config for BigTreeTech CB2/Pi2

[GitHub issue](https://github.com/armbian/build/labels/Task%2FTo-Do) reference: 
[Jira](https://armbian.atlassian.net/jira) reference number [AR-2332]

# How Has This Been Tested?

Build and load on BTT Pi 2, test peripherals and see how it runs for 5 minutes

# Checklist:

_Please delete options that are not relevant._

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules


[AR-2332]: https://armbian.atlassian.net/browse/AR-2332?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JohnTheCoolingFan commented:
> @JohnTheCoolingFan _please_ pick [armsurvivors@779b4ee](https://github.com/armsurvivors/armbian-build/commit/779b4eee5494d1f6da2c8691ea11af3a708652f8) -- you told me the other day I could drop it, but you seem to have forgotten?
> 
> @ColorfulRhino that addresses most of your comments, I think?

I did pick it, just didn't push onto a remote branch and it's not used in this branch. This is subject to change during the development of the PR. Pushed it to `btt-cb2-rpardini`. I have a number of branches for btt cb2 and cb1 and was organizing them yesterday, deleting unneeded ones. Will have to reevaluate branch this PR currently uses vs others I have.

rpardini commented:
> it's not used in this branch

Yeah, but reviews are costly -- please don't waste our time, it's precious. Thanks.

JohnTheCoolingFan commented:
> Yeah, but reviews are costly -- please don't waste our time, it's precious. Thanks.

When the PR is ready for review, I'll mark it as such. For now - it's a draft.

rpardini commented:
Well I sent https://github.com/armbian/build/pull/6874 (also draft) - I don't (or plan to) have the board

JohnTheCoolingFan commented:
Switched it to the @rpardini config. The built image does boot but there is no hdmi output, no ethernet connection and I still can't get proper input via UART pins. I don't have dedicated usb-to-uart converters yet and have been using a chinese arduino nano clone with a ch340 chip and the MCU reset pin tied to ground. I can get board's output butt can't input anything.

JohnTheCoolingFan commented:
Noticed that BTT had a custom `armbianEnv.txt` and there was `overlay=hdmi`, tried setting this in the image, didn't make the hdmi work.

I'm getting out of ideas. The device tree is same, based on the same kernel, so why is it not giving any output? The config is set to have output on both serial and display. And also Ethernet doesn't work even though it probably uses the same exact driver (based on 6.6 kernel).
Some help would really be appreciated. I can probably safely use an arduino to read serial output but not send until a proper usb-uart adapter comes in.

igorpecovnik commented:
@jernejsk Any ideas on HDMI trouble?

JohnTheCoolingFan commented:
Tried vendor's image, a minimal Armbian build. It had a lot of additional device tree overlays, including hdmi, which did get me a working hdmi output, but simply putting `overlays=hdmi` doesn't enable that output on the armbian build of this commit.
Also, I couldn't get any hardware UART from it even with a dedicated USB to UART adapter bases on a FT232RL chip. It didn't work on the vendor image and didn't work on the PR's commit build, even trying different UARTs and speeds (25000000, 15000000, 1500000, 115200, 9600).

igorpecovnik commented:
> but simply putting overlays=hdmi doesn't enable that output on the armbian build of this commit.

What if you put HDMI to device tree?

JohnTheCoolingFan commented:
It's already in the device tree: https://github.com/armbian/build/pull/6860/files#diff-d2416d7f0b4589c184941fa192bd7b1fd0baf0adbcf15b42f235be3fb728918cR262-R282

JohnTheCoolingFan commented:
Ok, I finally got some progress with getting a tty. I shouldn't have traced down the gpio pins from rk3566 datasheet, the tx/rx pins are the same as in Btt Pi (CB1) or CM4. The pins are not marked as uart in the official manual, which just makes it extremely confusing. I now have a fully functional tty and can debug problems further.

Alright, so now I would like to first set up ethernet so I don't have to use uart as often. In the current config, the board is having problems setting up ethernet:
```
[   23.648519] rk_gmac-dwmac fe010000.ethernet eth0: Register MEM_TYPE_PAGE_POOL RxQ-0
[   23.650924] rk_gmac-dwmac fe010000.ethernet eth0: PHY [stmmac-0:00] driver [YT8531 Gigabit Ethernet] (irq=POLL)
[   24.651122] rk_gmac-dwmac fe010000.ethernet: Failed to reset the dma
[   24.651145] rk_gmac-dwmac fe010000.ethernet eth0: stmmac_hw_setup: DMA engine initialization failed
[   24.651152] rk_gmac-dwmac fe010000.ethernet eth0: __stmmac_open: Hw setup failed
[   25.189408] rk_gmac-dwmac fe010000.ethernet eth0: Register MEM_TYPE_PAGE_POOL RxQ-0
[   25.191615] rk_gmac-dwmac fe010000.ethernet eth0: PHY [stmmac-0:00] driver [YT8531 Gigabit Ethernet] (irq=POLL)
[   26.195094] rk_gmac-dwmac fe010000.ethernet: Failed to reset the dma
[   26.195115] rk_gmac-dwmac fe010000.ethernet eth0: stmmac_hw_setup: DMA engine initialization failed
[   26.195122] rk_gmac-dwmac fe010000.ethernet eth0: __stmmac_open: Hw setup failed
```

JohnTheCoolingFan commented:
Putting the boot log gist here just for the record: https://gist.github.com/JohnTheCoolingFan/5df87cc4b87298d6bc21e046bfcff5f6

JohnTheCoolingFan commented:
Tried booting with the new changes, and I tried `overlays=hdmi` (I don't really know what it does but it seems the device tree has hdmi enabled?) on edge (6.10), no HDMI still and no wifi either. Didn't check for bluetooth, Ethernet doesn't work with the same error in kernel log, but both branches boot which is good.
I don't really know how to diagnose this further. @rpardini any ideas?

warasilapm commented:
Hey, just coming here to say I really appreciate the work you're putting into this, @JohnTheCoolingFan. I went looking for a way to compile an image for CB2 myself instead of relying on BTT's image and have been following your efforts for the last month or so. I'm not really familiar with this build process and don't really have much development bandwidth at the moment but if I do in the future I'll let you know. If I could help perform some tests or something less technical please let me know!

JohnTheCoolingFan commented:
Did some device tree adjustments I was hoping would make ethernet work, but no, unfortunately. HDMI still not working, wifi varies from boot to boot. Onboard eMMC works when enabled and the PCIe slot works with an NVMe drive. Didn't test sound.

JohnTheCoolingFan commented:
Ethernet is working now

JohnTheCoolingFan commented:
The board freezes after some time , reset button works but I don't know the cause

JohnTheCoolingFan commented:
Bluetooth seems to work, `bluetoothctl` recognizes the hardware and can use it, although while csanning to try and connect to my phone it froze again and I decided to stop. I'll mark bluetooth as working for now

JohnTheCoolingFan commented:
Temporarily using BigTreeTech u-boot as it fixes the freezing issue.

JohnTheCoolingFan commented:
Made the vendor device tree work on the board, unfortunately the freezing still occurs, so that means the defconfig is the deciding factor.
I now regret I didn't save the patches that had changes to the defconfig when I was trying to adapt the device tree. But I'll trace back my steps and try to adapt the vendor defconfig. Some things I do remember need changing are (after upgrading via `make oldconfig`): turning off kernel device tree usage, image fitting disable.

JohnTheCoolingFan commented:
Well, I adapted the vendor defconfig through `make oldconfig` but the result is still freezing. Which tells me that there might be something I wrongly turned switched while adapting the defconfig with generic-rk3568 on the other screen or that the reason BTT's `stable-4.19-cb2` u-boot works is deeper, potentially in one of the commits they used for their tree. Which I hope is not the case.

My patches are quite disorganized now, as it is easier to incrementally make a change and test it, then continue from what I had. It would've been nice if I didn't need to add on patches for each try and there were some way to combine patches and maybe choose the patches to apply and then include in the output patch. Once I do find a working configuration, I will put it into one patch and remove the other, rebasing to remove redundant commits as well.

JohnTheCoolingFan commented:
I've made multiple attempts to tweak the defconfig settings one way or another, but nothing seems to work. I suspect that one option that could be causing the issues is `CONFIG_SPL_BOARD_INIT`, which I cannot turn on in the latest u-boot due to a lack of a board file in `board/` dir. The option is enabled in the vendor's defconfig for 2017.09, yet there is no board tweaks file. I don't know how that feature works and what has changed.

I need help from people knowledgeable with u-boot, otherwise I doubt I can make any progress.

JohnTheCoolingFan commented:
There is a config `rk3568_defconfig` in the bigtreetech tree, and I assume it's a generic rk3568 defconfig. What's confusing is that taking a diff view shows that the bigtreetech config has barely any difference compared to `rk3568_defconfig`, and most of the changes are already enabled if it's hardware support or just don't matter it seems.
Which leads me to think that it's not just the config.
The problem is very weird and non-standard, I don't know a proper way to debug it, even though with help of some people I did narrow it down to u-boot, even though the problem happens when the kernel is running. 
I feel like I'm running in circles.

warasilapm commented:
> I did narrow it down to u-boot, even though the problem happens when the kernel is running.

I've only the most cursory understanding of u-boot. Is BTT using an older version or something with a bunch of custom patches on it for the CB1?

JohnTheCoolingFan commented:
BTT is using an older version (2017.09), potentially a different base branch. Other than that for CB2 they just added the device tree and defconfig. For CB1 they use a completely different branch, but in the same repo.

JohnTheCoolingFan commented:
Found the base for CB2 BigTreeTech u-boot: https://github.com/radxa/u-boot/tree/stable-4.19-rock3

This suggests that I may find some potential solutions by exploring radxa rock 3 series.

JohnTheCoolingFan commented:
I've decided to use bigtreetech u-boot for now if it works on current. Leaving mainline/kwiboo (will maybe follow-up with an actual mainline later) on edge. Doing test builds and boots to confirm behavior for both current and edge, will make the PR ready for review then.

JohnTheCoolingFan commented:
`current`: boots and works fine
`edge`: boots but freezes

JohnTheCoolingFan commented:
Fixed the typo and switched to mainline u-boot

JohnTheCoolingFan commented:
Rebased and added the missing patch, now building

JohnTheCoolingFan commented:
Same with mainline 2024.10
I already tried generic defconfig. Discussed the problem with @rpardini on discord, he thinks this might be a combination of both u-boot and kernel, something about voltage regulators.
For instance, here is one of the faulty boots: https://paste.next.armbian.com/yabuvohexa
And rpardini provided a fully working rk3568 boot log: https://paste.next.armbian.com/iquqiyusih

For now we've come to a conclusion that keeping vendor's u-boot for `current` branch while keeping mainline on `edge` and working on it more afterwards.

Kwiboo commented:
> Maybe @Kwiboo has some insights since he created the `generic-rk3568_defconfig`.

Took a quick peek at the u-boot patch, and can see that the kernel device tree does not match the u-boot device tree, so first I would change u-boot to use the more mainline friendly kernel device tree, add a `arch/arm/dts/rk3566-bigtreetech-cb2-u-boot.dtsi` with a `#include "rk356x-u-boot.dtsi"` and rename the defconfig to `bigtreetech-cb2-rk3566_defconfig`.

The defconfig is missing some options such as:
```
CONFIG_SPL_DM_SEQ_ALIAS=y
CONFIG_SYS_NS16550_MEM32=y
CONFIG_SYS_RELOC_GD_ENV_ADDR=y
CONFIG_DWC_ETH_QOS_ROCKCHIP=y
CONFIG_PHY_REALTEK=y
CONFIG_PHY_GIGE=y
CONFIG_LED=y
CONFIG_LED_GPIO=y
```
and contain some unneeded options such as:
```
CONFIG_SPL_TINY_MEMSET=y
CONFIG_DM_REGULATOR_GPIO=y
CONFIG_REGULATOR_PWM=y
CONFIG_PWM_ROCKCHIP=y
```

I also recommend to use the latest ddr init blob (`rk3566_ddr_1056MHz_v1.21.bin`) and bl31 (`rk3568_bl31_v1.44.elf`).

Finally I suggest you send a proper device tree to upstream Linux :-)

JohnTheCoolingFan commented:
Thanks! I've already sent an email to linux-rockchip with a patch for bigtreetech-cb2 device tree, used the kernel device tree from this PR. I don't really know what a "proper" device tree is, so I expect I'll get replies telling me what I should tweak. Bringing the DT to mainline is in progress.

As for syncing the device tree with the kernel one, I'm doing that bit by bit. For instance, commit [58f80dc](https://github.com/armbian/build/pull/6860/commits/58f80dce640277b6a2ca6c0f4e7f8f36ef70f827), but there is more to change and sync.

Thanks for pointing out the missing defconfig options that are missing or those that are unneeded, I will fix that. And will switch to fresh blobs too.

JohnTheCoolingFan commented:
Applied all of the @Kwiboo suggestions. The behavior hasn't changed, still freezing at same-ish time mark. Squashed the patches into one (still wishing the tooling would allow editing the same patch instead of having to create new ones). Still no clue why vendor u-boot version and config works while mainline doesn't and seemingly no tweaks help... I haven't tried doing radxa u-boot + patch to recreate btt's fork, just to eliminate some possibilities, but I doubt that's going to change anything. 7 years between releases, something might've changed in how a device tree is treated, or some other u-boot logic.

JohnTheCoolingFan commented:
small update tl;dr:
- `current` using vendor u-boot due to "magic gremlins" but otherwise board works fine (Ethernet, serial, nvme, emmc, etc)
- `edge` is left for experiments on trying to get mainline u-boot working.

JohnTheCoolingFan commented:
Rebased for rockchip64-6.12. I've also made attempts modifying u-boot or kernel device trees to fix the freezing issue. I tried changing voltage regulator definitions. Some resulted in error to boot, some made no difference at all.

I also found some peculiar details, or maybe even conflicts between the device trees:
1. Pin `gpio0 RK_PC5` is used for `vcc5v0_host` in the u-boot device tree but that pin is used as `snps,reset-gpio` in kernel device tree. The GPIO mode is different: `GPIO_ACTIVE_HIGH` in u-boot and `GPIO_ACTIVE_LOW` in kernel. For further tests, I could try disabling the `gmac1` node and see what happens.
2. Kernel device tree and u-boot device tree don't seem to share usb voltage regulators `pinctrl` at all: `0 RK_PC5` in u-boot and `0 RK_PB5`, `0 RK_PB6`, `4 RK_PC6` and `4 RK_PC4` in kernel.

These might be clues to voltage regulators being misconfigured in one or the other, or even gpio being an issue.

One of earlier hints of voltage regulators being misconfigured is from [dmesg](https://paste.next.armbian.com/yabuvohexa):
```
[    0.735069] debugfs: Directory 'vcc5v0_host' with parent 'regulator' already present!
[    0.735675] debugfs: Directory 'vcc5v0_host' with parent 'regulator' already present!
<...>
[    2.310654] rk808-regulator rk808-regulator.2.auto: there is no dvs0 gpio
[    2.311300] rk808-regulator rk808-regulator.2.auto: there is no dvs1 gpio
<...>
[   11.449100] edt_ft5x06 2-0038: supply vcc not found, using dummy regulator
[   11.449964] edt_ft5x06 2-0038: supply iovcc not found, using dummy regulator
[   11.450431] edt_ft5x06 2-0038: touchscreen probe failed
```
I am a bit confused on how problematic these messages are, especially the latter two. The first one seems to just be a conflict of... Something.
I'll also record the dmesg of a boot with bigtreetech u-boot to be able to compare against each other.

JohnTheCoolingFan commented:
While the freezing problem is annoying, there is still some stuff to fix. I've fixed a bunch of stuff reported in dmesg and attempted various tweaks to help with the problem of freezing.

New problems:
- Sometimes when rebooting I get a kernel panic: https://paste.next.armbian.com/pufehufega
- Non-fatal error that I don't know the meaning of in u-boot: `rockchip_rk3066_i2c i2c@fdd40000: pinctrl_select_state_full: uclass_get_device_by_phandle_id: err=-19`
- kernel errors about rk809_sound: 
  ```
  [   11.775546] asoc-simple-card rk809-sound: ASoC: Failed to add route MICBIAS1(*) -> Mic Jack(*)
  [   11.775572] asoc-simple-card rk809-sound: ASoC: Failed to add route Mic Jack(*) -> IN1R(*)
  [   11.775581] asoc-simple-card rk809-sound: ASoC: Failed to add route HPOL -> Headphone Jack(*)
  [   11.775590] asoc-simple-card rk809-sound: ASoC: Failed to add route HPOR -> Headphone Jack(*)
  ```
  
Fixed:
- `debugfs: Directory 'vcc5v0_host' with parent 'regulator' already present!` was fixed by giving unique `regulator-name` to each regulator
- `platform rk809-sound: deferred probe pending: asoc-simple-card: parse error` which I think also solves hdmi sound problem but I'm not sure.

I do admit I over-fixated on the critical problem and was ignoring the smaller issues that could be quickly investigated, fixed and checked. Although it might seem that not fixing the major problem is less effective, it is effective to spend some time on smaller issues because I can't find info on the major problem for a long time.

Maybe it's something about the u-boot loading the kernel and then something is trying to switch at 170 seconds. I don't have enough knowledge on u-boot and the loading process to know what it is, and research is slow. Bringing back the "help needed" label.

rpardini commented:
> * `current` using vendor u-boot due to "magic gremlins" but otherwise board works fine (Ethernet, serial, nvme, emmc, etc)
> * `edge` is left for experiments on trying to get mainline u-boot working.

I'd say having a `current` (with vendor u-boot and mainline kernel) that enables people to get going is good enough to merge, if you're willing to squash some/most of those 28 commits ;-) And yes you can keep on working on edge calmly.

It's really important to send the contributions upstream, as there are way more people who can help on the kernel/u-boot mailing lists.

JohnTheCoolingFan commented:
I've sent the current 6.6 kernel device tree to the linux-rockchip mailing list as a patch and subscribed to the mailing list. I did not receive any replies on it, but the patch did get posted, so I have no idea whether it will get accepted or not. I will continue improving the device tree, benefitting all branches, both `current` and `edge`. There is some stuff missing compared to the vendor's image based on 4.19. There are some features I can't test, like csi and dsi interfaces, but most of the stuff is not problematic. I will squash some of the patches and commits.

JohnTheCoolingFan commented:
Alright, a person on Discord messaged me about his usage of bigtreetech pi 2 6.11 with an image based on orangepi 3b. I've tried his dtb, and it works without the freezing issue. I will work on adding the proper dts.

JohnTheCoolingFan commented:
Oh, just realized adding the device tree files would be helpful: 
[rk3566-bigtreetech-pi2.dtb.gz](https://github.com/user-attachments/files/17794205/rk3566-bigtreetech-pi2.dtb.gz)
[rk3566-bigtreetech-pi2-by-Cowboy-full.dts.txt](https://github.com/user-attachments/files/17794207/rk3566-bigtreetech-pi2-by-Cowboy-full.dts.txt)

JohnTheCoolingFan commented:
With the diff I was provided, I could make a raw-ish device tree from a kernel 6.6 orangepi-3b device tree:
[rk3566-bigtreetech-pi2-by-Cowboy-from-diff.dts.txt](https://github.com/user-attachments/files/17837485/rk3566-bigtreetech-pi2-by-Cowboy-from-diff.dts.txt)
and [rk3566-orangepi-3b-full.dts.txt](https://github.com/user-attachments/files/17837486/rk3566-orangepi-3b-full.dts.txt) for reference

These device trees are pretty raw and as such are not suitable to be put into the `patch` folder I think, so I am working on a proper device tree, that is using the includes and labels instead of raw phandles.
I still had no good results from this, any help appreciated.

JohnTheCoolingFan commented:
Finally got around to making a more meaningful and human-readable "diff" of the patch, so that it can be used to use an existing orangepi 3b device tree and patched. No actual patch right now, just a human-readable summary. But I wish I could have some program that automatically resolves phandles at least.

- CPU voltage regulator: `syr837@40` (`reg = <0x40>`) --> `tsc4525@1c` (`reg = <0x1c>`)
- `pwm`:
    - `pwm0` `pinctrl-0`: `<&pwm0m0-pins>` --> `<&pwm0m1-pins>`
    - `pwm7`: disabled
    - `pwm12` `pinctrl-0`: `<&pwm12m0-pins>` --> `<&pwm12m1-pins>`
    - `pwm13` `pinctrl-0`: `<&pwm13m0-pins>` --> `<&pwm13m1-pins>`
    - `pwm14` `pinctrl-0`: `<&pwm14m0-pins>` --> `<&pwm14m1-pins>`
    - `pwm15` `pinctrl-0`: `<&pwm15m0-pins>` --> `<&pwm15m1-pins>`
- Ethernet `gmac1`:
    - `snps,reset-gpio`: `<&gpio3 RK_PC2 1>` --> `<&gpio0 RK_PC5 1>`
    - `reset-gpios`: <&gpio0 RK_PB6 0> --> `<&gpio1 RK_PB2 0>`
- `sdmmc1`:
    - Added `rockchip,default-sample-phase = <0x5a>` (90 in decimal)
    - Added `sdio-wifi@1`
- `sfc`: removed `flash@0`
- `i2c`:
    - `i2c2` `pinctrl-0`: `<&i2c2m0-xfer>` --> `<&i2c2m1-xfer>`
    - `i2c3`: Added `tft_tp: ns2009@48`
- `spi`:
    - `spi1` `pinctrl-0`: `<&spi1m0-cs0 &spi1m0-cs1 &spi1m0-pins>` --> `<&spi1m1-cs0 &spi1m1-pins>`
    - `spi1`: added `can_mcp2515: can-mcp2515@0`
    - `spi1`: added `tft_35: st7789v@1`
    - `spi3` `pinctrl-0`: `<&spi3m0-cs0 &spi3m0-cs1 &spi3m0-pins>` --> `<&spi3m1-cs0 &spi3m1-pins>`
- `uart`:
    - `uart1`: added `dma-names`
    - `uart1`: added `bluetooth`
    - `uart5` `pinctrl-0`: `<&uart5m0-xfer>` --> `<&uart5m1-xfer>`
    - `uart7` `pinctrl-0`: `<&uart7m0-xfer>` --> `<&uart7m2-xfer>`
- `usb`:
    - `usb2phy0/host-port` `phy-supply`: `<&vbus>` --> `<&vcc5vc0-host-regulator &vcc5v0-org-regulator>`
    - `usb2phy0/host-port` `phy-supply`: `<&vcc5v0_usb2t &vcc5v0_usb2b>`
- `pinctrl`:
    - Added `bt`:
        - `bt_enable: bt-enable-h`: `<2 RK_PB7 0 &pcfg-pull-down>`
        - `bt_host_wake: bt-host-wake-l`: `<2 RK_PC1 0 &pcfg-pull-down>`
        - `bt_wake: bt-wake-l`: `<2 RK_PC0 0 &pcfg-pull-none>`
    - `sdio-pwrseq`: added `wifi_host_wake: wifi-host-wake-l`: `<2 RK_PB1 0 &pcfg-pull-down>`
    - `usb`:
        - `vcc5v0-host-en`: `<0 RK_PA6 0 &pcfg-pull-none>` --> `<0 RK_PB6 0 &pcfg-pull-none>`
        - `vcc5v0-otg-en`: `<0 RK_PA5 0 &pcfg-pull-none>` --> `<0 RK_PB5 0 &pcfg-pull-none>`
        - added `vcc5v0_usb2t_en: vcc5v0_usb2t-en`: `<3 RK_PD5 0 &pcfg-pull-none>`
        - added `vcc5v0_usb2b_en: vcc5v0-usb2b-en`: `<4 RK_PC4 0 &pcfg-pull-none>`
    - `work-led`:
        - `leds-gpio`: `<0 RK_PC0 0 &pcfg-pull-none>` --> `<0 RK_PB7 0 &pcfg-pull-none>`
        - added `blue_led: led-blue`: `<4 RK_PA1 0 &pcfg-pull-none>`
    - `pcie/pcie-drv`: `<0 RK_PB7 0 &pcfg-pull-none>` --> `<4 RK_PB1 0 &pcfg-pull-none>`
    - added `mcp2515_int_pin` with `mcp2515_int_pin: mcp2515_in_pin`: `<4 RK_PA3 0 &pcfg-pull-up>`
- `leds`:
    - original node's `pinctrl-0` and `pinctrl-names` moved to `led@1`
    - added `led@0`:
        - `gpios`: `<&gpio4 RK_PA1 1>`
        - `label`: `"power_led`
        - `pinctrl-0`: `<&blue_led>`
        - `pinctrl-names`: `"default`
        - `linux,default-trigger`: `default-on`
    - `led@1`:
        - `gpios`: `<&gpio0 RK_PC0 0>` --> `<&gpio0 RK_PB7 1>`
        - `pinctrl-0` and `pinctrl-names` of `leds` were moved here.
- removed `sprd-mtty`
- removed `uwe-bsp`
- regulators:
    - `vcc3v3-pcie-regulator` `gpio`: `<&gpio0 RK_PB7 0>` --> `<&gpio4 RK_PB1 0>`
    - `vcc5v0-host-regulator` `gpio`: `<&gpio0 RK_PA6 0>` --> `<&gpio0 RK_PB6 0>`
    - `vcc5v0-host-regulator` `regulator-name`: `"vcc5v0_host"` --> `"vcc5v0_host3"`
    - `vcc5v0-org-regulator` `gpio`: `<&gpio0 RK_A5 0>` --> `<&gpio0 RK_PB5 0>`
    - `vcc5v0-org-regulator` `regulator-name`: `"vcc5v0_otg"` --> `vcc5v0_otg3"`
    - added `vcc5v0_usb2t: vcc5v0-usb2t-regulator`:
        - `pinctrl-0`: `<&vcc5v0_usb2t_en>`
        - `gpio`: `<&gpio3 RK_PD5 0>`
        - `regulator-name`: `"vcc5v0_usb2t"`
        - otherwise inherit `vcc5v0-otg-regulator`
    - added `vcc5v0_usb2b: vcc5v0-usb2b-regulator`
        - `pinctrl-0`: `<&vcc5v0_usb2b_en>`
        - `gpio`: `<&gpio4 RK_C4 0>`
        - `regulator-name`: `"vcc5v0_usb2b"`
        - otherwise inherit `vcc5v0-otg-regulator`
    - added `ext_cam_clk: ext-cam-clk`:
        - `clock-frequency`: `<25000000>`
        - `clock-output-names`: `ext_cam_clk`
        - `#clock-cells`: `<0>`
    - added `can_mcp2515_osc: can-mcp2515-osc`:
        - `clock-frequency`: `<8000000>`
        - `#clock-cells`: `<0>`

I will now try to make a new device tree, based on orangepi-3b and inserted through the patches system.

JohnTheCoolingFan commented:
Yeah, I have a fully working device now. dmesg: https://paste.next.armbian.com/ofumurekuy

Will commit my changes and push, then can get to re-organizing and probably remove the tree that is currently in this PR at all.

JohnTheCoolingFan commented:
For some reason the edge (6.12) won't output anything in the same serial port I've used before, maybe my serial/screen setup is messed up again, I don't know. I did change the compatible string for the tsc4525, but I don't even get the u-boot output, weird. I'll focus on current (6.6) for now.

JohnTheCoolingFan commented:
> For some reason the edge (6.12) won't output anything in the same serial port I've used before, maybe my serial/screen setup is messed up again
Nevermind, forgot to put the sd card in

rpardini reviewed (approved):
Dear John, what a journey. Congratulations. Merge at will.

Prokopiev commented:
Great work! Can I ask you if hdmi is working? Tried to build your branch and no luck with output

JohnTheCoolingFan commented:
HDMI works for me on all kernel versions.

rpardini commented:
> Oops, forgot to specify the board maintainer as myself.

Nah, that doesn't work like that, unfortunately. Igor has some secret script that updates this based on some database, somewhere.

JohnTheCoolingFan commented:
Alright fixed the mistake of misreading a diff and putting `reset-gpios` into gmac1, which was supposed to go to pcie2x1. That's why pcie didn't work right away. Updated my message with the summary of differences between opi3b and btt-cb2

JohnTheCoolingFan commented:
@rpardini could you please update the review / make a new approval so that this can be merged?

rpardini reviewed (approved):
sure!

bigtreetech commented:
@JohnTheCoolingFan You are what happens when a wizard and a legend have a baby. Please send an email to luke@bigtree-tech.com so that I can get in touch with you.

feldgendler commented:
@JohnTheCoolingFan this is amazing, thank you so much for this work!

Now that the right configuration is merged, what needs to be done to set up regular automatic builds for this platform?

anarsoul commented:
> HDMI works for me on all kernel versions.

It works only with certain resolutions. I get 800x480 and 640x480 modes rejects with `MODE_NOCLOCK`, so it doesn't work with BTT HDMI5

raizer88 commented:
Hi @JohnTheCoolingFan thank you for your work. 
Do you know if the merge you made is used by BTT for their beta or it's different? 
https://github.com/bigtreetech/CB2/releases/tag/V2.0.0 
I have a new pi2 and I want to understand which build to install.

JohnTheCoolingFan commented:
@raizer88 they are using kernel 6.1. they might have used the uboot part of this PR, but I doubt they used anything for teh kernel, it's probably still different.

anarsoul commented:
I've got BTT HDMI5 working with my CB2, see https://github.com/armbian/build/pull/7970
Other screens (e.g. BTT HDMI7) would require adding more PLL rates

anarsoul commented:
@JohnTheCoolingFan on my CB2 CPU regulator is tcs,tcs4526, not 4525. Could you confirm that it is indeed 4525 on your board?

4526 has a different ID so the driver will refuse to probe it as 4525, but other than that it is compatible. Without vdd_cpu cpufreq won't work.

@bigtreetech to confirm CPU regulator model on CB2.

anarsoul commented:
> @JohnTheCoolingFan on my CB2 CPU regulator is tcs,tcs4526, not 4525. Could you confirm that it is indeed 4525 on your board?
> 
> 4526 has a different ID so the driver will refuse to probe it as 4525, but other than that it is compatible. Without vdd_cpu cpufreq won't work.
> 
> @bigtreetech to confirm CPU regulator model on CB2.

Nevermind, looks like it's something broken on 6.14-rc with regulators.

JohnTheCoolingFan commented:
I was told about an error with the cpu voltage regulator:
```
[   87.232764] cpu cpu0: _set_opp_voltage: failed to set voltage (1150000 1150000 1150000 mV): -6
[   87.232823] cpu cpu0: Failed to set regulator voltages: -6
[   87.232849] cpufreq: __target_index: Failed to change cpu frequency: -6
```
I can confirm that on edge 6.14 from main, but changing the device tree value to `tcs4526` doesn't make the error go away.
@anarsoul is this your issue or is it something else?
Also, I checked btt's source code, they use `tcs4525`, so with these I am confident it's `tcs4525`, not `tcs4526`

anarsoul commented:
> I was told about an error with the cpu voltage regulator:
> 
> ```
> [   87.232764] cpu cpu0: _set_opp_voltage: failed to set voltage (1150000 1150000 1150000 mV): -6
> [   87.232823] cpu cpu0: Failed to set regulator voltages: -6
> [   87.232849] cpufreq: __target_index: Failed to change cpu frequency: -6
> ```
> 
> I can confirm that on edge 6.14 from main, but changing the device tree value to `tcs4526` doesn't make the error go away. @anarsoul is this your issue or is it something else? Also, I checked btt's source code, they use `tcs4525`, so with these I am confident it's `tcs4525`, not `tcs4526`

Yeah, this is the issue I'm seeing on 6.14.

On 6.12 it was a different issue with wrong compatible string (you merged my fix earlier today). I checked fan53555 driver code and it doesn't really care if it's tcs4525 or tcs4526 (since they are compatible), but the one used in my CB2 is tcs4526 according to the chip ID.

I can debug the issue on 6.14 in coming days unless you are looking into it.

anarsoul commented:
BTW, it looks like all regulators are broken on 6.14, you can check `cat /sys/kernel/debug/regulator/regulator_summary` output on 6.14 to see that it cannot read current voltages, while it is fine on 6.12.

JohnTheCoolingFan commented:
I see... Well, good to know it's not because of something device-specific.
As for the regulator model, I didn't remember at the time of writing that I merged that earlier. Will probably need to follow that up with a patch for the mainline device tree as well, will do that sometime later.

anarsoul commented:
Mainline device tree is fine, I compared it with one from Armbian yesterday night. I'm pretty sure that 6.12 would boot with mainline dtb, although I didn't test it.

JohnTheCoolingFan commented:
But mainline is using `tcs4525` as the compatibility string... I'm confused.

anarsoul commented:
Sorry if I confused you. It doesn't really matter if dts specifies tcs4525 or tcs4526, the driver would accept both and will work fine. The regulators are compatible at register level, they just have different chip IDs, but the driver doesn't care which one it is as long as its either tcs4525 or tcs4526.

JohnTheCoolingFan commented:
Thank you for the clarification.

anarsoul commented:
@adelyser nice catch! Are you planning to submit the fix upstream?

adelyser commented:
I have to backtrack here, that seemed to work, but after several power cycles, it's apparently not the fix. If I power cycle the device, it does eventually work properly, which leads me to think there is some sort of issue in the order that drivers are loading things or a race condition that is out of sync.

anarsoul commented:
@adelyser can you share `dmesg` for working and non-working cases? You may need to increase printk buffer size, just pass `log_buf_len=16M` to kernel arguments, e.g. by appending it to `extraargs` in `/boot/armbianEnv.txt`
